lib/apollo-federation/entity.rb

```diff
@@ -9,5 +9,18 @@ class Entity < GraphQL::Schema::Union
     def self.resolve_type(object, context)
       context[object]
     end
+
+    # The main issue here is the fact that an union in GraphQL can't be an interface according
+    # to the [spec](https://spec.graphql.org/October2021/#sec-Unions.Type-Validation), but at
+    # the same time, according to the Federation spec, an interface can be an Entity, and an Entity
+    # is an union. Therefore, we have to extend this validation to allow interfaces as possible types.
+    def self.assert_valid_union_member(type_defn)
+      if type_defn.is_a?(Module) &&
+         type_defn.included_modules.include?(ApolloFederation::Interface)
+        # It's an interface entity, defined as a module
+      else
+        super(type_defn)
+      end
+    end
```

Comment:
I totally get why we would do this but I was curious if you think we could tweak the implementation to avoid creating a union that is technically invalid according to the GraphQL spec.

I was thinking we keep the requirement that all entity interface implementations are entities, and continue to check that when getting the list of possible types for the entities field, but then drop it from the final list of possible types and then in the entities resolver we use the `resolve_type` method on an interface and just force that resolution to happen.

Thinking something like this, though I haven't tested to see how well this plays with GraphQL ruby's `after_lazy`.

```ruby
      maybe_lazies = grouped_references_with_indices.map do |typename, references_with_indices|
        references = references_with_indices.map(&:first)
        indices = references_with_indices.map(&:last)

        # TODO: Use warden or schema?
        type = context.warden.get_type(typename)
        if type.nil? || !(type.kind == GraphQL::TypeKinds::OBJECT || (type.kind == GraphQL::TypeKinds::Interface && has_key_directive(type)))
          raise "The _entities resolver tried to load an entity for type \"#{typename}\"," \
                ' but no entity type of that name was found in the schema'
        end
        
        # ... EXISTING CODE calling resolve reference(s) ...
        
           final_result[i] = context.schema.after_lazy(result) do |resolved_value|
              # force the type to the concrete type
              type_to_set = if type.kind == GraphQL::TypeKinds::Interface
                type_class.resolve_type(resolved_value, context)
              else
                type
              context[resolved_value] = type_to_set
              resolved_value
            end
```

Do you think something like that would work?